art/attacks/evasion/auto_conjugate_gradient.py

```diff
@@ -0,0 +1,714 @@
+
+        if self.estimator.nb_classes == 2 and y.shape[1] == 1:
+            raise ValueError(
+                "This attack has not yet been tested for binary classification with a single output classifier."
+            )
+
+        x_adv = x.astype(ART_NUMPY_DTYPE)
+
+        for _ in trange(max(1, self.nb_random_init), desc="ACG - restart", disable=not self.verbose):
+            # Determine correctly predicted samples
+            y_pred = self.estimator.predict(x_adv)
+            if self.targeted:
+                sample_is_robust = np.argmax(y_pred, axis=1) != np.argmax(y, axis=1)
+            elif not self.targeted:
+                sample_is_robust = np.argmax(y_pred, axis=1) == np.argmax(y, axis=1)
+
+            if np.sum(sample_is_robust) == 0:
+                break
+
+            x_robust = x_adv[sample_is_robust]
+            y_robust = y[sample_is_robust]
+            x_init = x[sample_is_robust]
+
+            n = x_robust.shape[0]
+            m = np.prod(x_robust.shape[1:]).item()
+            random_perturbation = (
+                random_sphere(n, m, self.eps, self.norm).reshape(x_robust.shape).astype(ART_NUMPY_DTYPE)
+            )
+
+            x_robust = x_robust + random_perturbation
+
+            if self.estimator.clip_values is not None:
+                clip_min, clip_max = self.estimator.clip_values
+                x_robust = np.clip(x_robust, clip_min, clip_max)
+
+            perturbation = projection(x_robust - x_init, self.eps, self.norm)
+            x_robust = x_init + perturbation
+
+            # Compute perturbation with implicit batching
+            for batch_id in trange(
+                int(np.ceil(x_robust.shape[0] / float(self.batch_size))),
+                desc="ACG - batch",
+                leave=False,
+                disable=not self.verbose,
+            ):
+                batch_index_1, batch_index_2 = batch_id * self.batch_size, (batch_id + 1) * self.batch_size
+                x_k = x_robust[batch_index_1:batch_index_2].astype(ART_NUMPY_DTYPE)
+                x_init_batch = x_init[batch_index_1:batch_index_2].astype(ART_NUMPY_DTYPE)
+                y_batch = y_robust[batch_index_1:batch_index_2]
+
+                p_0 = 0
+                p_1 = 0.22
+                var_w = [p_0, p_1]
+
+                while True:
+                    p_j_p_1 = var_w[-1] + max(var_w[-1] - var_w[-2] - 0.03, 0.06)
+                    if p_j_p_1 > 1:
+                        break
+                    var_w.append(p_j_p_1)
+
+                var_w = [math.ceil(p * self.max_iter) for p in var_w]
+
+                # self.eta = np.full((self.batch_size, 1, 1, 1), 2 * self.eps_step).astype(ART_NUMPY_DTYPE)
+                bs = x_k.shape[0]
+                eta = np.full((bs, 1, 1, 1), self.eps_step).astype(ART_NUMPY_DTYPE)
+                self.count_condition_1 = np.zeros(shape=(bs,))
+                gradk_1 = None
+                sk_1 = None
+                sk = None
```

Comment:
Would it be possible to change the name `sk` into a name with at least 3 charadters?